### PR TITLE
Rename one_plus_dlnf_dlnc to thermodynamic_factor

### LINKED ISSUE #2726: Rename `one_plus_dlnf_dlnc` to `thermodynamic_factor`
Maybe for a separate (small) issue, should we rename this to `thermodynamic_factor`? It is not much longer, and more informative.

_Originally posted by @brosaplanella in https://github.com/pybamm-team/PyBaMM/pull/2419#discussion_r1116150328_

### PR DESCRIPTION
**Pull request by jeromtom**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #2726

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [X] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [X] No style issues: `$ pre-commit run` (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [X] All tests pass: `$ python run-tests.py --all`
- [X] The documentation builds: `$ python run-tests.py --doctest`

You can run unit and doctests together at once, using `$ python run-tests.py --quick`.

## Further checks:

- [X] Code is commented, particularly in hard-to-understand areas
- [X] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by jeromtom** [comment]
`1 + dlnf/dlnc` is referenced multiple times and cannot be replaced with `Thermodynamic factor` manually. I am in the process of creating a script to do the same.

**Comment by tinosulzer** [comment]
Use this https://stackoverflow.com/questions/37346481/how-do-i-find-and-replace-all-occurrences-in-all-files-in-visual-studio-code

**Comment by jeromtom** [comment]
Surely, will try that. Made the changes running `for /f "delims=" %f in (C:\Users\jerto\Desktop\Replacer\filelist.txt) do powershell -Command "(Get-Content \"%f\") -replace '1 \+ dlnf/dlnc', 'Thermodynamic factor' | Set-Content \"%f\"" ` in cmd

**Comment by jeromtom** [comment]
That too after using `findstr` which @Saransh-cpp had introduced to me, to find references and store them into filelist.txt.

**Comment by jeromtom** [comment]
image
This was way easier.

**Comment by tinosulzer** [comment]
You have to click on the cross which takes you to codecov then see which line is not covered and fix bugs or add tests as appropriate so that the line is covered. You should test this locally e.g. by temporarily adding `print("here")` next to the line that should be covered, and making sure you can see "here" in the command line when you run the test. There is a bug in your current changes to `parameter_values.py`, but I won't tell you what since you should be able to figure it out yourself (and it's good to practice finding bugs)

**Comment by tinosulzer** [comment]
Are you even testing this locally like I suggested or are you using trial and error with the GitHub tests. Please test locally to make sure your fix works.

**Comment by jeromtom** [comment]
I haven't been able to do the [developer install](https://pybamm.readthedocs.io/en/latest/source/user_guide/installation/install-from-source.html) to run tests locally. Will be doing it by the books.

**Comment by tinosulzer** [comment]
The installation instructions are a bit confusing (hence the GSOC project to improve them) but you should be able to just go to the pybamm root repository where you downloaded it, create a virtual environment via pip or conda, and then do `pip install -e .`

Local installation is a required prerequisite for participating in GSOC

**Comment by jeromtom** [comment]
Thanks, @tinosulzer. I will do it soon.

**Comment by jeromtom** [comment]
Local installation is complete. Working on writing tests now.

**Comment by ayeankit** [comment]
Is this still open? I would like to work on it.

**Comment by jeromtom** [comment]
Hey @ayeankit, I am very close to finishing this PR.

**Comment by jeromtom** [comment]
@tinosulzer , can you please help me out here? I have done the local installation.

File: pybamm/parameters/parameter_values.py [hunk 1]
```diff
@@ -317,6 +317,12 @@ def update(self, values, check_conflict=False, check_already_exists=True, path="
                         + "sure you want to update this parameter, use "
                         + "param.update({{name: value}}, check_already_exists=False)"
                     )
+            # specific check for depreciated parameter "1 + dlnf/dlnc"
+            if name in self.keys() == "1 + dlnf/dlnc":
+                raise ValueError(
+                    "parameter '{}' is depreciated"
+                    + "and has been changed to `Thermodynamic factor`"
+                )
```
Comment (tinosulzer) [inline pybamm/parameters/parameter_values.py:325]: You need to change this to
```python3
if name == "1 + dlnf/dlnc":
                raise ValueError(
                    "parameter '1 + dlnf/dlnc' has been renamed to "
                    "'Thermodynamic factor'"
                )
```
and also add a test for it

**Comment by jeromtom** [comment]
@tinosulzer there are two merge conflicts here. 
Kindly review.

**Comment by tinosulzer** [comment]
You have to merge the develop branch and fix the conflicts.

File: tests/unit/test_parameters/test_parameter_values.py [hunk 2]
```diff
@@ -159,6 +159,20 @@ def test_check_parameter_values(self):
                 {"Negative electrode LAM constant propotional term": 1}
             )
 
+    def test_parameter_deprecation(self):
```
Comment (tinosulzer) [inline tests/unit/test_parameters/test_parameter_values.py:162]: This test should test that `parameter_values.update({"1 + dlnf/dlnc": 1})` errors. Currently it's not testing anything useful. Also, use `assertRaisesRegex`

**Comment by tinosulzer** [comment]
It will be much more efficient to test your changes locally to make sure coverage passes, instead of relying on the CI, as I explained how to do in this comment https://github.com/pybamm-team/PyBaMM/pull/2727#issuecomment-1443790945

**Comment by jeromtom** [comment]
I haven't understood how the print statements can be use for testing coverage. Can we start using pytest or Coverage.py or Pytest-cov like mentioned [in this website](https://www.sealights.io/agile-testing/test-metrics/python-code-coverage/) in the future? Because then we can properly run coverage tests locally.

**Comment by tinosulzer** [comment]
You can already use coverage.py locally if you want. That's what the CI uses. But that still requires running all the unit tests, which takes a few minutes. It's much faster to run the individual test files (`python path/to/test.py`) and check the command-line output. Once you've added `print("here")` in the right place, then "here" should appear in the command line output if that line is being covered by that particular test. Otherwise, it shows you that the line isn't being covered.

**Comment by jeromtom** [comment]
BTW I had done local tests using `python run-tests.py --unit` but it did not include coverage.

**Comment by jeromtom** [comment]
Is the following placement of the print statement correct?
image

**Comment by jeromtom** [comment]
image
@tinosulzer, I now totally agree with temporarily using print statements. It is way faster.

**Review by tinosulzer — APPROVED** [review]

**Comment by tinosulzer** [comment]
Thanks, will merge once final test passes

**Comment by jeromtom** [comment]
This is the first test code that I have ever written. @tinosulzer, thank you for your guidance.

**Comment by tinosulzer** [comment]
Congrats on writing your first test!

**Comment by tinosulzer** [comment]
@all-contributors add @jeromtom for code, tests

**Pull request merged**